core/corerepo/gc.go

```diff
@@ -89,46 +86,70 @@ func GarbageCollect(n *core.IpfsNode, ctx context.Context) error {
 	if err != nil {
 		return err
 	}
-	rmed, err := gc.GC(ctx, n.Blockstore, n.DAG, n.Pinning, roots)
-	if err != nil {
-		return err
-	}
+	rmed := gc.GC(ctx, n.Blockstore, n.DAG, n.Pinning, roots)
+
+	return CollectResult(ctx, rmed, nil)
+}
 
+func CollectResult(ctx context.Context, gcOut <-chan gc.Result, cb func(*cid.Cid)) error {
+	var errors []error
+loop:
 	for {
 		select {
-		case _, ok := <-rmed:
+		case res, ok := <-gcOut:
 			if !ok {
-				return nil
+				break loop
+			}
+			if res.Error != nil {
+				errors = append(errors, res.Error)
+			} else if res.KeyRemoved != nil && cb != nil {
+				cb(res.KeyRemoved)
 			}
 		case <-ctx.Done():
-			return ctx.Err()
+			errors = append(errors, ctx.Err())
+			break loop
 		}
 	}
 
+	switch len(errors) {
+	case 0:
+		return nil
+	case 1:
+		return errors[0]
+	default:
+		return NewMultiError(errors...)
+	}
 }
 
-func GarbageCollectAsync(n *core.IpfsNode, ctx context.Context) (<-chan *KeyRemoved, error) {
-	roots, err := BestEffortRoots(n.FilesRoot)
-	if err != nil {
-		return nil, err
+func NewMultiError(errs ...error) *MultiError {
+	return &MultiError{errs[:len(errs)-1], errs[len(errs)-1]}
+}
+
+type MultiError struct {
```

Comment:
Mentioning that it implements the error interface is pretty redundant and not standard go convention